src/main/scala/lms/thirdparty/cuda/cuda.scala

```diff
@@ -47,9 +47,202 @@ trait CudaOps extends Base with SizeTOps with StackArrayOps with CLibs with Cuda
+
+  // this CUDA kernel clips the abstract value of the cuda array
+  def cudaArrayClipAt[T:Manifest](res: Rep[Array[T]], bound: Rep[T], size: Rep[Int]) =
+    cudaFunction[Unit]("clipAt", basic_config, Unwrap(res), Unwrap(bound), Unwrap(size))(Seq(0), Seq(0), Set[Int]())
+
+  abstract class Dim3
+  def dim3(a: Rep[Int], b: Rep[Int] = unit(1), c: Rep[Int] = unit(1)): Rep[Dim3] =
+    libFunction("dim3", Unwrap(a), Unwrap(b), Unwrap(c))(Seq[Int](), Seq[Int](), Set[Int](), Backend.UNSAFE)
+
+
+  // How do we generate the kernels (instead of manually writing them)
+  // the cuda functions need Dim3 typed inputs :)
+  def cudaGlobalFun[A:Manifest, B:Manifest](f: Rep[A] => Rep[B]) =
+    Wrap[(A,Dim3,Dim3)=>B](__topFun(f, 1, xn => Unwrap(f(Wrap[A](xn(0)))), "__global__"))
+
+  def cudaGlobalFun[A:Manifest,B:Manifest,C:Manifest](f: (Rep[A], Rep[B]) => Rep[C])=
+    Wrap[(A,B,Dim3,Dim3)=>C](__topFun(f, 2, xn => Unwrap(f(Wrap[A](xn(0)), Wrap[B](xn(1)))), "__global__"))
+
+  def cudaGlobalFun[A:Manifest,B:Manifest,C:Manifest,D:Manifest](f: (Rep[A], Rep[B], Rep[C]) => Rep[D]) =
+    Wrap[(A,B,C,Dim3,Dim3)=>D](__topFun(f, 3, xn => Unwrap(f(Wrap[A](xn(0)), Wrap[B](xn(1)), Wrap[C](xn(2)))), "__global__"))
+
+  def cudaGlobalFun[A:Manifest,B:Manifest,C:Manifest,D:Manifest,E:Manifest](f: (Rep[A], Rep[B], Rep[C], Rep[D]) => Rep[E]) =
+    Wrap[(A,B,C,D,Dim3,Dim3)=>E](__topFun(f, 4, xn => Unwrap(f(Wrap[A](xn(0)), Wrap[B](xn(1)), Wrap[C](xn(2)), Wrap[D](xn(3)))), "__global__"))
+
+  def cudaGlobalFun[A:Manifest,B:Manifest,C:Manifest,D:Manifest,E:Manifest,F:Manifest](f: (Rep[A], Rep[B], Rep[C], Rep[D], Rep[E]) => Rep[F]) =
+    Wrap[(A,B,C,D,E,Dim3,Dim3)=>F](__topFun(f, 5, xn => Unwrap(f(Wrap[A](xn(0)), Wrap[B](xn(1)), Wrap[C](xn(2)), Wrap[D](xn(3)), Wrap[E](xn(4)))), "__global__"))
+
+  // When coding kernel functions, we often need some kernel variables
+  def gridDimX = cmacro[Int]("gridDim.x")
+  def gridDimY = cmacro[Int]("gridDim.y")
+  def blockDimX = cmacro[Int]("blockDim.x")
+  def blockDimY = cmacro[Int]("blockDim.y")
+  def blockDimZ = cmacro[Int]("blockDim.z")
+
+  def blockIdxX = cmacro[Int]("blockIdx.x")
+  def blockIdxY = cmacro[Int]("blockIdx.y")
+  def threadIdxX = cmacro[Int]("threadIdx.x")
+  def threadIdxY = cmacro[Int]("threadIdx.y")
+  def threadIdxZ = cmacro[Int]("threadIdx.z")
+
 }
 
-trait CCodeGenCudaOps extends CCodeGenSizeTOps with CCodeGenLibs {
+trait CCodeGenCudaOps extends CCodeGenSizeTOps with CudaCodeGenLibFunction with CCodeGenLibs {
   // need to register the headers
   registerHeader("<cuda_header.h>")
+
+  override def remap(m: Manifest[_]): String = m.runtimeClass.getName match {
+    case s: String if s.endsWith("Dim3") => "dim3"
+    case _ => super.remap(m)
+  }
+
+  override def shallow(n: Node): Unit = n match {
+    case n @ Node(s,"@", (f:Backend.Exp)::args, _) if ((graphCache(f.asInstanceOf[Sym]) match {
```

Comment:
@Kraks we cannot use `Adapter.g.Def` in code gen because it is reset to `null` before  code gen

we can use `graphCache` though, which is a copy of `Adapter.g.graphCache`